File: 25. Reverse Nodes in k-Group/25. Reverse Nodes in k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if (!head || k == 1) return head;
        
        ListNode dummy(0);
        dummy.next = head;
        ListNode* prev = &dummy;
        ListNode* curr = head;
        ListNode* next = nullptr;
        
        int count = 0;
        for (ListNode* temp = head; temp; temp = temp->next)
            count++;
        
        while (count >= k) {
            curr = prev->next;
            next = curr->next;
            for (int i = 1; i < k; i++) {
                curr->next = next->next;
                next->next = prev->next;
                prev->next = next;
                next = curr->next;
            }
            prev = curr;
            count -= k;
        }
        
        return dummy.next;
    }
};

File: 25. Reverse Nodes in k-Group/25. Reverse Nodes in k-Group.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? reverseKGroup(ListNode? head, int k) {
    if (head == null || k == 1) return head;
    
    ListNode dummy = ListNode(0);
    dummy.next = head;
    ListNode? prev = dummy;
    ListNode? curr = head;
    ListNode? next;
    
    int count = 0;
    for (ListNode? temp = head; temp != null; temp = temp.next)
      count++;
    
    while (count >= k) {
      curr = prev?.next;
      next = curr?.next;
      for (int i = 1; i < k; i++) {
        curr?.next = next?.next;
        next?.next = prev?.next;
        prev?.next = next;
        next = curr?.next;
      }
      prev = curr;
      count -= k;
    }
    
    return dummy.next;
  }
}